src/lib/statistics/column_statistics.cpp

```diff
@@ -0,0 +1,444 @@
+      // distinction between integers and floats
+      // for integers "< value" means that the new max is value <= value - 1
+      // for floats "< value" means that the new max is value <= value - ε
+      if (std::is_integral_v<ColumnDataType>) {
+        return estimate_range(_min, value - 1);
+      }
+      // intentionally no break
+      // if ColumnType is a floating point number, OpLessThanEquals behaviour is expected instead of OpLessThan
+      [[fallthrough]];
+    }
+    case PredicateCondition::LessThanEquals:
+      return estimate_range(_min, value);
+
+    case PredicateCondition::GreaterThan: {
+      // distinction between integers and floats
+      // for integers "> value" means that the new min value is >= value + 1
+      // for floats "> value" means that the new min value is >= value + ε
+      if (std::is_integral_v<ColumnDataType>) {
+        return estimate_range(value + 1, _max);
+      }
+      // intentionally no break
+      // if ColumnType is a floating point number,
+      // OpGreaterThanEquals behaviour is expected instead of OpGreaterThan
+      [[fallthrough]];
+    }
+    case PredicateCondition::GreaterThanEquals:
+      return estimate_range(value, _max);
+
+    case PredicateCondition::Between: {
+      DebugAssert(static_cast<bool>(value2), "Operator BETWEEN should get two parameters, second is missing!");
+      auto casted_value2 = type_cast<ColumnDataType>(*value2);
+      return estimate_range(value, casted_value2);
+    }
+
+    default:
+      Fail("Estimation not implemented for request PredicateCondition");
+  }
+}
+
+/**
+ * Specialization for strings as they cannot be used in subtractions.
+ */
+template <>
+ColumnValueEstimate ColumnStatistics<std::string>::estimate_predicate_with_value(
+    const PredicateCondition predicate_condition, const AllTypeVariant& variant_value,
+    const std::optional<AllTypeVariant>& value2) const {
+  // if column has no distinct values, it can only have null values which cannot be selected with this predicate
+  if (distinct_count() == 0) {
+    return {0.f, without_null_values()};
+  }
+
+  auto casted_value = type_cast<std::string>(variant_value);
+  switch (predicate_condition) {
+    case PredicateCondition::Equals: {
+      return estimate_equals_with_value(casted_value);
+    }
+    case PredicateCondition::NotEquals: {
+      return estimate_not_equals_with_value(casted_value);
+    }
+    // TODO(anybody) implement other table-scan operators for string.
+    default: { return {non_null_value_ratio(), without_null_values()}; }
+  }
+}
+
+template <typename ColumnDataType>
+ColumnValueEstimate ColumnStatistics<ColumnDataType>::estimate_predicate_with_value_placeholder(
+    const PredicateCondition predicate_condition, const ValuePlaceholder& value,
+    const std::optional<AllTypeVariant>& value2) const {
+  switch (predicate_condition) {
+    // Simply assume the value will be in (_min, _max) and pick _min as the representative
```

Comment:
I don't understand this.

Reply:
Didn't change this. I think its trying to say it makes the statistical estimation under the assumption that the placeholder has the same value as _min, lacking better alternatives.